File: swift/isbn-verifier/Sources/IsbnVerifier/IsbnVerifier.swift
import Foundation

func isValid(_ input: String) -> Bool {
    var isbn = input
    isbn.removeAll { !($0.isWholeNumber || $0 == "X") }

    // validate 
    let regex = try! NSRegularExpression(pattern: "^\\d{9}[\\dX]$")
    let range = NSRange(location: 0, length: isbn.count)
    guard 1 == regex.numberOfMatches(in: isbn, range: range)
          else {return false}

    // split into digits
    var digits = isbn.prefix(9).compactMap { $0.wholeNumberValue! }
    // add check digit: "X" == 10
    digits.append(isbn.last!.wholeNumberValue ?? 10)

    let isbnSum = digits
        .enumerated()
        .map {(i, d) in d * (10 - i)}
        .reduce(0, +)
    return isbnSum % 11 == 0
}

File: euphoria/isbn-verifier/isbn-verifier.ex
include std/sequence.e

public function is_valid(sequence isbn)
    isbn = remove_all('-', isbn)

    if length(isbn) != 10 then
        return 0
    end if

    if isbn[$] = 'X' then
        isbn[$] = 10 + '0'
    end if
    isbn -= '0'

    integer sum = 0
    for i = 1 to length(isbn) - 1 do
        if isbn[1] < 0 or isbn[i] > 9 then
            return 0
        end if
        sum += isbn[i] * (11 - i)
    end for
    sum += isbn[10]

    return remainder(sum, 11) = 0
end function
